### ISSUE TITLE
[RFC] Hierarchical Model Averaging (Hierarchical SGD)

### ISSUE THREAD

**Issue by wayi1** [pytorch/pytorch#71325]
## Infeasibility of Frequent Global Synchronization
Nowadays powerful industry users (not necessarily in FAANG) like Cruise, Microsoft, and Tesla are moving towards scaling large-scale training to 100+ or even 1K+ GPUs. DDP, as the most well-adopted data parallelism implementation in PyTorch, fully synchronizes the gradients of all the processes at each iteration by default, which may lead to a low scaling efficiency in this case (let's assume that it may not be feasible to afford high-bandwidth network environment like InfiniBand/EFA or high-end GPUs for many companies).

The key is that, the global synchronization at each iteration can become infeasible at a very large scale, at least for some industry users. Therefore, we may need to consider replacing the default SyncSGD with a combination of frequent partial synchronization and infrequent global synchronization. One solution I contributed to PyTorch is [post-localSGD](https://github.com/pytorch/pytorch/issues/59699), based on model averaging paradigm, as implemented in this [module](https://github.com/pytorch/pytorch/tree/master/torch/distributed/algorithms/model_averaging). Now it seems that the ownership is transferred to @rohan-varma.

## Hierarchical Model Averaging
One effective approach is **hierarchical model averaging**, namely **hierarchical SGD** ([ICLR'2021](https://arxiv.org/pdf/2007.13819.pdf), [AAAI'2022](https://arxiv.org/pdf/2010.12998.pdf)), which can be viewed as an extended version of local SGD. Local SGD essentially shows a two-level hierarchy. An oversimplified example is that, given 32 GPUs on 8 machines, every 4 GPUs on the same machine is organized as a sub process group, 8 subgroups in total. Each subgroup runs allreduce at every iteration. In the meantime, a global allreduce is executed every 8 or 16 iterations. Additionally, some other tricks are needed to ensure convergence efficiency.

If powerful users plan to scale to at least hundreds of GPUs, a two-level hierarchy may not be high enough to support infrequent global synchronizations. Therefore, we can consider executing allreduce in a more hierarchical way, which can be user-defined. For example, we can imagine a 4-level hierarchy like this:

- Level I: 4 GPUs on the same machine sync every iteration;
- Level II: 32 GPUs sync every 8 iterations;
- Level III: 128 GPUs sync every 16 iterations;
- Level IV: 512 GPUs sync every 32 iterations.

To avoid the potential data race, at each iteration, there will be at most one synchronization, which should be at the highest level that triggers allreduce.

Note that recently Facebook has published a relevant [paper](https://dl.acm.org/doi/abs/10.1145/3447548.3467084) on KDD'21, although it is deployed on CPU clusters, so I guess this feature (built as a generic PyTorch module) can also be beneficial for Facebook itself.

## Benefits

1. Improving the scaling efficiency, especially if the model is relatively large and can cause non-trivial data transfer cost.
2. Smoothing out potentially large computation variance and hence mitigating stragglers, based on the assumption that stragglers occur randomly among all the processes at each iteration. The mitigation effect is similar to gradient accumulation, but hierarchical model averaging still runs optimizer step at every iteration, it can often give a higher accuracy than a simple gradient accumulation implementation.

### Alternatives

As mentioned above, gradient accumulation can be an alternate approach. However, there can be a few disadvantages:
1. Usually the # of accumulated steps before each synchronization cannot be too large due to the potential model quality regression.
2. Gradient accumulation still runs global allreduce, which can be less communication-efficient.

### Additional context

This can be considered as a DDP feature request. To facilitate the triage:

Labels: feature, module:ddp, oncall:distributed.
CC PyTorch Distributed DDP developers: @mrshenli @pritamdamania87 @zhaojuanmao @rohan-varma

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by pritamdamania87** [comment]
Thanks for the suggestion @wayi1! I was wondering if our current DDP communication hooks are generic enough for you to try out these different ideas and implement them on your end in the mean time? It does make sense to have native support in PyTorch for some of these ideas, but we would like to primarily make DDP hackable/pluggable such that ideas like these can be prototyped easily by users themselves and then we can incorporate some of these techniques as builtin features into DDP itself.

I guess since this idea also involves model averaging, we might need a generic pluggable model averaging component as well here?

**Comment by wayi1** [comment]
> I was wondering if our current DDP communication hooks are generic enough for you to try out these different ideas and implement them on your end in the mean time?

@pritamdamania87 I don't expect PT-D to allocate anyone to implement this in the near term. Actually I plan to work on the implementation at Cruise.

IMO, the current DDP should be already flexible enough for supporting this idea -- in other words, I don't request any enablement work from DDP.

> I guess since this idea also involves model averaging, we might need a generic pluggable model averaging component as well here?

We will see if that is necessary. Will need some code reviews from PT-D when iterating over the implementation, in order to better incorporate into PyTorch.

**Comment by vadimkantorov** [comment]
Maybe related: https://twitter.com/m_ryabinin/status/1481994243761713153

**Comment by wayi1** [comment]
> Maybe related: https://twitter.com/m_ryabinin/status/1481994243761713153

Thanks for the pointer! It's an interesting read.

I think this paper focuses more on a fault-tolerant allreduce implementation over extremely heterogeneous and extremely unreliable (potentially **preemptible**) devices. Specifically, the heterogeneous experimental setup are:
1) 81 heterogeneous GPUs across 64 servers and across two data centers, and **additional latency is injected**. 
2) 66 **preemptible** heterogeneous GPUs, and 34 of them are various devices rented on a public marketplace, spread across 3 continents.

Honestly, I don't think the above extreme heterogeneity and unreliability would be practical for almost all the decent industry users. By contrast, this RFC targets a much more common scenario in industry -- a large number of homogeneous GPUs: each machine usually has multiple GPUs (so the # of machines are much smaller than the # of GPUs), and these machines are usually in the same cluster. I really doubt Moshpit AllReduce could still outperform more commonly used model averaging approaches or Gossip SGD.

**Comment by vadimkantorov** [comment]
Yeah, I mostly referred to its separation of nodes into groups and aggregation happening within those groups, so maybe that aspect is related to the proposed hierarchical grouping / aggregation.

**Comment by rohan-varma** [comment]
Thanks for the proposal Yi! I wonder if this can be incorporated into the model averaging library, or if we can directly use ModelAverager/PeriodicModelAverager to implement this idea (both of the interfaces look quite generic and pluggable): https://github.com/pytorch/pytorch/tree/master/torch/distributed/algorithms/model_averaging

**Comment by wayi1** [comment]
> Yeah, I mostly referred to its separation of nodes into groups and aggregation happening within those groups, so maybe that aspect is related to the proposed hierarchical grouping / aggregation.

Thanks @vadimkantorov! Both ideas are indeed relevant in this sense.

**Comment by wayi1** [comment]
> Thanks for the proposal Yi! I wonder if this can be incorporated into the model averaging library, or if we can directly use ModelAverager/PeriodicModelAverager to implement this idea (both of the interfaces look quite generic and pluggable): https://github.com/pytorch/pytorch/tree/master/torch/distributed/algorithms/model_averaging

Yes, I am working on prototyping this idea based on the existing model averaging foundation.
